### PR TITLE
#29938 changes for Turkish input issue

### PR DESCRIPTION
**Pull request by alatas**
I changed discussed lines, added missing tests, and synced tests between C# and VB for Turkish character issue #29938

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
One thing missing from these tests is actual cases where symbols contain different values of all those interesting turkish-`i`s.  We need to have tests for those, along with different expected ways that turkish-speaking users would want to write their code, while ensuring that intellisense properly picks accordingly.

@alatas I'm depending on you for this.  You're the most qualified person to indicate what would be the desired/expected behavior from a turkish-speaking user.  Can you write some tests that exercise these sorts of symbol+completion interactions as well?

Thanks!

**Comment by CyrusNajmabadi** [comment]
@gafter I definitely agree that invariant is likely *not* the right culture to use.  That said, there is a lot of trickyness in this space, because intellisense works in a fuzzy manner, that isn't necessarily equivalent to how the language do *strict* matching.  For example, users *may* want to write an english letter that matches a similar foreign version (or vice versa).  In these cases, CaseInsensitiveComparison may not consider those values equal.  But users would expect them to match and complete (fixing the letter to the actual correct letter from that API).

**Comment by CyrusNajmabadi** [comment]
Note: this is why, in general, our code uses an intentional mix of both "current culture" and "en-us culture".  This is because there is an inherent problem in determining intent given that APIs are nearly universally english/latin, but current culture is very specific and changes from machine to machine and user to user.  

We've found we need to understand current-culture to respect the user's culture (especially for APIs written in their home language).  However, we also need to understand en-us, otherwise their current culture can cause problems when trying to match against things like standard latin APIs.  I would expect a fix to likely need some combination of both cultures.

**Comment by alatas** [comment]
@CyrusNajmabadi 
> One thing missing from these tests is actual cases where symbols contain different values of all those interesting turkish-`i`s. We need to have tests for those, along with different expected ways that turkish-speaking users would want to write their code, while ensuring that intellisense properly picks accordingly.

I actually tried to accomplish this. 

I added "TestMatchWithTurkishIWorkaround4" includes class named `IFADE` (means statement in Turkish). It's written as `ifade` in lowercase, or `İFADE` in uppercase in Turkish texts. But when we name a class, we choose `IFADE` or `Ifade`, because the English alphabet is more suitable and common to name a class. However, when code uses current culture to make comparisons, It handles this as a Turkish word and doesn't match correctly. 

This test also handles `if` problem (especially in VB). I type `if` when everytime I need to write `If` in VB, and then I type a space. CurrentCulture settings matches and shows `If` in window somehow but not to prioritize it. When I type `if`, intellisense shows `If`, `#If` etc. But when I typed a space It selects `IFADE` instead of `If`. This is a unexpected behavior. It have to select `If` because of its priority.

In one of my projects, There is a class named `ISLEM_TARIFE`. When I typed `if` and a space, It selects this class but not `If`. I curiosly examined this very unexpected behavior and found out that, It shows `If`, `#If` and `ISLEM_TARIFE` but not prioritize the `If` and select the `ISLEM_TARIFE` because this word has `IF` after 10th character.

I tried to check very similar problem with the new "TestMatchWithTurkishIWorkaround3" also. It includes a class named `TARIFE` (means tariff in Turkish). It's written as `tarife` or `TARİFE` in Turkish texts. But, when we create a class, we choose English alphabet and name class as `TARIFE`. But, as a native Turkish speaker, I wrote `Tarife` to find that class. Again, when code uses current culture, it won't match `TARIFE` class with `Tarife` keystrokes. Because they are so different in Turkish.

---

> We also need tests where there is a mix of these things. i.e. `is` spelled with normal `i` as well as versions with the different turkish `I`s. We'd then want to validate the right thing is selected depending on what sort of `i` the user entered.

In other words, Turkish `i` is not different from ordinary `i`. Because, we write statements, class names, variable names etc. in English. Using abnormal `İ` in these particular areas is not common and is not suitable. So, if code matches as en-US culture, it won't be a problem I think.

When, I changed the to Invariant Culture, I tested whole intellisense code base and didn't see a problem. But, I understand your concerns about InvariantCulture. you may consider use en-US in general term for comparisons.

---

Overall, I think you only need these unit tests but not the fix for culture related codes. I can close this pull request and open a new one with tests only, if you want. 

And also, I can add some new tests expected and unexpected behaviors in Turkish.

**Comment by CyrusNajmabadi** [comment]
>  Using abnormal İ in these particular areas is not common and is not suitable. So, if code matches as en-US culture, it won't be a problem I think.

I'm not sure i agree there.  We've made these changes in the past because we got reports for non-english-speaking language devs who *do* use native-language words in their symbols.  I've seen first hand source code bases that use these types of symbols.  So i don't think we can arbitrarily change behavior here if that would regress things for those codebases.

**Comment by CyrusNajmabadi** [comment]
> And also, I can add some new tests expected and unexpected behaviors in Turkish.

That would definitely be great!

**Comment by alatas** [comment]
@CyrusNajmabadi I updated the commit and this pull request, and I removed the code changes, added 16 new unit test in total to C# and VB, side by side. These tests cover all different usage of `i`, `ı`, `İ` and `I` in a Turkish perspective. If you need more unit test or more information, please don't hesitate to ask.

**Comment by jinujoseph** [comment]
@alatas thanks for the contribution.
Will you be able to retarget this change to master branch.

**Comment by ivanbasov** [comment]
Thank you, @alatas ! I have created a PR borrowing your tests and adding a fix: https://github.com/dotnet/roslyn/pull/30391
